drivers: adc: adc_mcux_adc12: enable support for channel>=16 on S32K14X

marcin-wierzbicki opened the pull request:
Enable support for ADC on s32k1 platform.
Enable ADC on UCANS32K1SIC eval board.

Enable support for channels >= 16 on devices S32K146 and S32K148. Channels >= 16 are encoded as 100000b..101111b in SC1[ADCH] field.

We are getting everything ready to support ADC on the S32K146/8 (https://github.com/zephyrproject-rtos/hal_nxp/pull/407).

marcin-wierzbicki committed:
drivers: adc: adc_mcux_adc12: enable support for channel>=16 on S32K14X

Enable support for channels >= 16 on devices S32K146 and S32K148.
Channels >= 16 are encoded as 100000b..101111b in SC1[ADCH] field.

Signed-off-by: Marcin Wierzbicki <marcin.wierzbicki@accenture.com>

manuargue reviewed (changes requested):
There's no ADC support for S32K1 in upstream, so this patch cannot be tested until said support is added. Please add devicetree definitions for ADC nodes and enable it in tests/samples. This should work in the `ucans32k1sic` board, which is the only board supported at the moment.

Also update the `west.yaml` manifest to point to https://github.com/zephyrproject-rtos/hal_nxp/pull/407

marcin-wierzbicki committed:
soc: arm: nxp_s32: s32k1: add support for ADC

Add support for the Analog-to-Digital Converter (ADC).

Signed-off-by: Marcin Wierzbicki <marcin.wierzbicki@accenture.com>

marcin-wierzbicki committed:
board: ucans32k1sic: Enable ADC on UCANS32K1SIC evb

Enable ADC on the UCANS32K1SIC

Signed-off-by: Marcin Wierzbicki <marcin.wierzbicki@accenture.com>

marcin-wierzbicki committed:
samples: adc: Add overlay for UCANS32K1SIC

New file with ADC configuration for UCANS32K1SIC evb.
Add ucans32k1sic_s32k146 to platform_allow.

Signed-off-by: Marcin Wierzbicki <marcin.wierzbicki@accenture.com>

manuargue commented on samples/drivers/adc/adc_sequence/sample.yaml:
```diff
@@ -11,6 +11,7 @@ tests:
       - nrf52840dk/nrf52840
       - nrf54l15pdk/nrf54l15/cpuapp
       - nrf54h20dk/nrf54h20/cpuapp
+      - ucans32k1sic_s32k146
```
```suggestion
      - ucans32k1sic
```

manuargue commented on samples/drivers/adc/adc_dt/sample.yaml:
```diff
@@ -33,6 +33,7 @@ tests:
       - longan_nano/gd32vf103/lite
       - rd_rw612_bga
       - frdm_mcxn947/mcxn947/cpu0
+      - ucans32k1sic_s32k146
```
```suggestion
      - ucans32k1sic
```

manuargue replied:
pls update this one, otherwise twister will not detect the board

marcin-wierzbicki replied:
Hi, it seems that Twister isn't detecting the board in either case:

ERROR   - samples/drivers/adc/adc_dt/sample.drivers.adc.adc_dt - platform_allow - unrecognized platform - ucans32k1sic/s32k146

ERROR   - samples/drivers/adc/adc_dt/sample.drivers.adc.adc_dt - platform_allow - unrecognized platform - ucans32k1sic_s32k146

manuargue replied:
ok, use `ucans32k1sic` then. That's what I've been using in other tests/samples

manuargue replied:
ok, use `ucans32k1sic` then. That's what I've been using in other tests/samples

manuargue commented on boards/nxp/ucans32k1sic/ucans32k1sic-pinctrl.dtsi:
```diff
@@ -74,4 +74,44 @@
+
+	adc0_default: adc0_default {
+		group0 {
+			pinmux = <ADC0_SE0_PTA0>,
+				<ADC0_SE1_PTA1>,
+				<ADC0_SE2_PTA6>,
+				<ADC0_SE3_PTA7>,
+				<ADC0_SE4_PTB0>,
+				<ADC0_SE5_PTB1>,
+				<ADC0_SE6_PTB2>,
+				<ADC0_SE7_PTB3>,
+				<ADC0_SE8_PTB13>,
+				<ADC0_SE8_PTC0>,
+				<ADC0_SE9_PTC1>,
+				<ADC0_SE10_PTC2>,
+				<ADC0_SE11_PTC3>,
+				<ADC0_SE12_PTC14>,
+				<ADC0_SE13_PTC15>,
+				<ADC0_SE14_PTC16>,
+				<ADC0_SE15_PTC17>;
+			drive-strength = "low";
+		};
+	};
+
+	adc1_default: adc1_default {
+		group0 {
+			pinmux = <ADC1_SE0_PTA2>,
+				<ADC1_SE1_PTA3>,
+				<ADC1_SE14_PTB0>,
+				<ADC1_SE15_PTB1>,
+				<ADC1_SE7_PTB12>,
+				<ADC1_SE8_PTB13>,
+				<ADC1_SE2_PTD2>,
+				<ADC1_SE3_PTD3>,
+				<ADC1_SE6_PTD4>,
+				<ADC1_SE10_PTE2>,
+				<ADC1_SE11_PTE6>;
```
I don't see any ADC dedicated pin in the UCANS32K1SIC board schematics, hence there should not be default ADC pinmuxing for this board. But I'll let @PetervdPerk-NXP or @bperseghetti comment on this

igalloway replied:
![image](https://github.com/user-attachments/assets/34557b33-4e59-4c45-99c3-b742993314b2)

![image](https://github.com/user-attachments/assets/667f1569-094d-46dd-99ae-12d62f700e6c)

PTE0 - Need to determine if there is an ADC channel that can mux here (i don't think so?)

marcin-wierzbicki replied:
> I don't see any ADC dedicated pin in the UCANS32K1SIC board schematics, so AFAICT there should not be default ADC pinmuxing for this board. But I'll let @PetervdPerk-NXP or @bperseghetti comment on this

Hi, I’m looking for guidance on how to proceed. I understand that there aren’t dedicated ADC pins, but it’s possible to enable analog input on certain pins using alternate functions. Could it be that there’s just a minimal set of pins, such as PTC14 on ADC0, that can be used for this purpose?

manuargue replied:
Board dts should assign pin functions according to the schematic, so if there's none assigned on this board, I don't see the purpose of having them defined. The user can always extend the pinmuxing by application dt overlays for those IOs that can be used with ADC functions

manuargue commented:
@marcin-wierzbicki it seems that the conversion from raw to mv is not working? this is the output of `samples/drivers/adc/adc_dt/` when SW3 is pressed. Please let me know how you tested these changes so I can reproduce it on my side
```
ADC reading[28]:
- adc@4003b000, channel 12: 0 = 0 mV
ADC reading[29]:
- adc@4003b000, channel 12: 4095 = 0 mV
```

manuargue commented:
> I tested that the same way as you. Also noticed that conversion is not working.

well that may indicate a problem with your changes on this PR? I haven't had time to debug it yet, but I'd expect that's working.

marcin-wierzbicki commented:
> > I tested that the same way as you. Also noticed that conversion is not working.
> 
> well that may indicate a problem with your changes on this PR? I haven't had time to debug it yet, but I'd expect that's working.

Ok. Will check that. BTW is there any automatic testing that these samples need be integrated?

manuargue commented:
> > > I tested that the same way as you. Also noticed that conversion is not working.
> > 
> > 
> > well that may indicate a problem with your changes on this PR? I haven't had time to debug it yet, but I'd expect that's working.
> 
> Ok. Will check that. BTW is there any automatic testing that these samples need be integrated?

The samples, as well as tests, can be run with twister, and they usually include a harness that check expected output in the serial console. Besides the sample you enabled for the ucansick1 board, there is also https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/adc/adc_api, if you can also enable it would be great to have a minimal check that adc is working as expected on this board.

marcin-wierzbicki committed:
tests: adc: Add overlay for UCANS32K1SIC

New file with ADC configuration for UCANS32K1SIC evb.

Signed-off-by: Marcin Wierzbicki <marcin.wierzbicki@accenture.com>

manuargue commented on boards/nxp/ucans32k1sic/ucans32k1sic.dts:
```diff
@@ -192,3 +192,19 @@
 };
+
+&adc0 {
+	interrupts = <39 0>;
+	clk-source = <0>;
+	clk-divider = <1>;
```
move `interrupts`, `clk-source`, `clk-divider` to the soc dts, since it's not board configuration dependent

manuargue commented on drivers/adc/adc_mcux_adc12.c:
```diff
@@ -253,6 +264,9 @@ static const struct adc_driver_api mcux_adc12_driver_api = {
 	.read_async = mcux_adc12_read_async,
 #endif
+#if defined(CONFIG_SOC_SERIES_S32K1)
+	.ref_internal = ADC_S32K1_REF_INTERNAL_MV,
```
I propose to do this in a more generic way by defining a dt prop as other bindings do, e.g.
```
  vref-mv:
    type: int
    description: Indicates the reference voltage of the ADC in mV.
```
and then set it in the `ucans32k1sic` board's overlay as `vref-mv: <5000>;`. It is better to keep the drivers decoupled as much as possible from SoC or board specific values.

marcin-wierzbicki replied:
Good point. However, it appears that the adc_mcux_adc12 driver only supports reference = ADC_REF_INTERNAL. This means that the conversion function (adc_raw_to_millivolts_dt) uses ref_internal instead of vref_mv.

manuargue replied:
that's correct, the value of `vref-mv` can be used to set `api.ref_internal`, see for instance `adc_npcx.c`, `adc_nxp_s32_adc_sar.c` or `adc_rpi_pico.c` for inpiration

manuargue reviewed (changes requested):
some final comments, the rest LGTM

marcin-wierzbicki committed:
drivers: adc: adc_mcux_adc12: add API internal reference

Extend the adc_mcux_adc12 adc api to set the ref_internal to
a value of vref-mv property.

Signed-off-by: Marcin Wierzbicki <marcin.wierzbicki@accenture.com>

marcin-wierzbicki committed:
drivers: adc: adc_mcux_adc12: set the vref as a property

Add the vref reference voltage in the DTS, so that the adc driver
can retrieve the value for conversion.

Signed-off-by: Marcin Wierzbicki <marcin.wierzbicki@accenture.com>

marcin-wierzbicki committed:
boards: nxp: frdm_ke17z and frdm_ke17z512: add vref-mv value

Add vref value of 3300mV.

Signed-off-by: Marcin Wierzbicki <marcin.wierzbicki@accenture.com>

marcin-wierzbicki committed:
boards: nxp: twr_ke18f: add vref-mv value

Add vref value of 3300mV.

Signed-off-by: Marcin Wierzbicki <marcin.wierzbicki@accenture.com>

marcin-wierzbicki commented on boards/nxp/ucans32k1sic/ucans32k1sic.dts:
```diff
@@ -192,3 +192,15 @@
+
+&adc0 {
+	sample-time = <65>;
+	vref-mv = <5000>;
```
Yep. my bad :)

manuargue commented:
hi @anangl , could you pls have a look to this pr? thanks

manuargue commented:
hi @marcin-wierzbicki, please update the manifest to point hal_nxp to 466000a80e6eb5bbcb691bae936ec5654a7796d4

manuargue commented:
> hi @anangl , could you pls have a look at this pr? thanks

ping @decsny collaborator for ADC area, as I believe @anangl may be unavailable this time, would you mind having a look and possibly re-assigning this to you? Thanks

carlescufi commented:
@marcin-wierzbicki please rebase

marcin-wierzbicki commented:
@manuargue Thank you for your insightful feedback and assistance :)
